### ISSUE TITLE
Make the interpreter paths initialization more consistent

### ISSUE THREAD

**Issue by FFY00** [python/cpython#98947]
As Steve pointed out in https://github.com/python/cpython/issues/98790#issuecomment-1297864613, it would be really nice to make paths initialization more consistent across platforms.

I am creating this issue to start documenting all the quirks and inconsistencies.

-  #98790

**Comment by davidhewitt** [comment]
One quirk I have noticed seems to be that launching an embedded interpreter with `Py_Initialize` behaves differently with respect to virtualenvs depending on OS.

(Specifically here I'm talking about Rust applications embedding Python via PyO3 - which basically just calls `Py_InitializeEx` under the hood & doesn't try to do anything smart.)

On Linux, I've never had any problems - launching an embedded interpreter with a virtualenv loaded will access the packages from that environment.

On macOS and Windows, users have reported that the system Python appears to always be preferred, and while there's been some discussion about possible causes / solutions I wonder if you think it's relevant to this issue?
https://github.com/PyO3/pyo3/issues/1741
https://github.com/PyO3/pyo3/issues/1896

**Comment by FFY00** [comment]
Yep, I think that's definitely relevant to this issue. Just beware that I have no estimate of when this work might be completed, so I'd advise to still try to work things out downstream if these issues are affecting your users.

**Comment by davidhewitt** [comment]
Appreciate that, thanks. I think even if this work landed in 3.12, given we support all supported Python versions in PyO3 we probably will want to find some solution downstream before Python 3.11 hits EOL ;)

**Comment by zooba** [comment]
There's a nice long list of quirks marked in getpath.py already. I don't know whether they can all be resolved, or if they should.

Many differences are due to `/` being inherently secure on POSIX and inherently insecure on Windows (so e.g., we don't want to search up through all directories because the top level is a free-for-all).

For building alternate Python entry points[^1] that recognise venv, you will really want to look at getpath.py for the logic (search `venv`). Chances are it's not working because you're calling any one of the functions that override normal search, since an embedded Python shouldn't be using a venv.[^2]

[^1]: I call it this because `venv` is a feature of the `pythonX.Y` command, not the language. So once you replace the command, if you want it to behave the same, you're actually building an entrypoint. A hosted/embedded Python runtime should have its own set of library directories, rather than trying to rely on the discovery mechanisms used by `pythonX.Y`.

[^2]: See the previous footnote.
